File: app/src/main/java/com/example/purepawapp/ui/common/ViewModelFactory.java
package com.example.purepawapp.ui.common;

import androidx.annotation.NonNull;
import androidx.lifecycle.ViewModel;
import androidx.lifecycle.ViewModelProvider;

public class ViewModelFactory<T extends ViewModel> implements ViewModelProvider.Factory {

    public interface Creator<T extends ViewModel> {
        T create();
    }

    private final Creator<T> creator;

    public ViewModelFactory(Creator<T> creator) {
        this.creator = creator;
    }

    @NonNull
    @Override
    @SuppressWarnings("unchecked")
    public <U extends ViewModel> U create(@NonNull Class<U> modelClass) {
        return (U) creator.create();
    }
}

File: app/src/main/java/com/example/purepawapp/ui/common/ViewModelFactory.kt
package com.example.purepawapp.ui.common

import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider

class ViewModelFactory<T : ViewModel>(private val creator: () -> T) : ViewModelProvider.Factory {
    @Suppress("UNCHECKED_CAST")
    override fun <U : ViewModel> create(modelClass: Class<U>): U = creator() as U
}
